src/env.cc

```diff
@@ -299,6 +372,145 @@ void Environment::CreateProperties() {
+    if (!snapshot_data->ReadContextIndependentObject<TypeName>(               \
+            SnapshotReadData::kAllowEmpty).To(&field)) {                      \
+      return;                                                                 \
+    }                                                                         \
+    PropertyName##_.Reset(isolate_, field);                                   \
+  } while (0);
+  ENVIRONMENT_STRONG_PERSISTENT_TEMPLATES(V)
+#undef V
+
+#define V(PropertyName, TypeName)                                             \
+  do {                                                                        \
+    Local<TypeName> field;                                                    \
+    if (!snapshot_data->ReadObject<TypeName>(context,                         \
+            SnapshotReadData::kAllowEmpty).To(&field)) {                      \
+      return;                                                                 \
+    }                                                                         \
+    PropertyName##_.Reset(isolate_, field);                                   \
+  } while (0);
+  ENVIRONMENT_STRONG_PERSISTENT_VALUES(V)
+#undef V
+
+  uint64_t expected_native_module_count;
+  if (!snapshot_data->ReadUint64().To(&expected_native_module_count)) return;
+  for (uint64_t i = 0; i < expected_native_module_count; i++) {
+    std::string str;
+    if (!snapshot_data->ReadString().To(&str)) return;
+    native_modules_in_snapshot.insert(str);
+  }
+
+  uint64_t expected_base_object_count;
+  if (!snapshot_data->ReadUint64().To(&expected_base_object_count)) return;
+
+#ifdef DEBUG
+  std::set<const BaseObject*> deserialized_objects;
+#endif
+  for (uint64_t i = 0; i < expected_base_object_count; i++) {
+    std::string name;
+    if (!snapshot_data->StartReadEntry(nullptr).To(&name)) return;
+    BaseObject* obj = BaseObjectDeserializer::Deserialize(
+        name, this, snapshot_data);
+    if (obj == nullptr) return;
+#ifdef DEBUG
+    deserialized_objects.insert(obj);
+#endif
+  }
+
+#ifdef DEBUG
+  // Make sure all BaseObject instances have been attached to this Environment.
+  ForEachBaseObject([&](const BaseObject* obj) {
+    deserialized_objects.erase(obj);
+  });
+  CHECK(deserialized_objects.empty());
+#endif
+
+  Local<Context> ctx;
+  if (!snapshot_data->ReadObject<Context>(context).To(&ctx)) return;
+  if (ctx != context)
+    snapshot_data->add_error("Context mismatch");
+  snapshot_data->EndReadEntry();
+}
+
+void Environment::Serialize(SnapshotCreateData* snapshot_data) const {
+  snapshot_data->StartWriteEntry("Environment");
+  async_hooks()->Serialize(snapshot_data);
+  immediate_info()->Serialize(snapshot_data);
+  tick_info()->Serialize(snapshot_data);
+  stream_base_state_.Serialize(snapshot_data);
+  should_abort_on_uncaught_toggle_.Serialize(snapshot_data);
+  performance_state()->Serialize(snapshot_data);
+
+  snapshot_data->WriteBool(has_run_bootstrapping_code());
```

Comment:
Why do we need to serialize this? Isn't that predefined by how we serialize/deserialize the snapshot anyway?

Reply:
It currently is, but I don’t see any harm in writing/reading this. I can remove it if you like, as you mentioned, the situation is pretty straightforward at the moment.

Reply:
I guess the downside is just one extra bit? But from my understanding this is not dynamic, but more like an invariant we have in the code for internal CHECKs against bugs, so I can't see how useful this is unless we want to also serialize another Environment that's not bootstrapped in parallel (but what for? Whether something is bootstrapped in JS code or in C++ is rather arbitrary)

Reply:
@joyeecheung Yeah, I’ve removed this from the snapshot data. :+1: